app/src/views/GrantRoundDetails.vue

```diff
@@ -0,0 +1,254 @@
+import { BigNumber, Contract, ContractTransaction } from 'src/utils/ethers';
+import { daysAgo, formatAddress, isValidAddress, isValidUrl } from 'src/utils/utils';
+// --- Types ---
+import { GrantRound } from '@dgrants/types';
+import { GrantRound as GrantRoundContract } from '@dgrants/contracts';
+import { BigNumberish, utils } from 'ethers';
+
+function useGrantRoundDetail() {
+  const { grantRounds, poll } = useDataStore();
+  const { signer, userAddress } = useWalletStore();
+  const route = useRoute();
+
+  // get a single grantRound or an empty/error object (TODO: should the typings be modified to account for an empty object?)
+  const grantRound = computed(() => {
+    if (grantRounds.value) {
+      // filter for a matching GrantRound
+      const round = grantRounds.value.filter((round) => round.address == route.params.address);
+
+      return <GrantRound>(round.length ? round[0] : { error: `No GrantRound @ ${route.params.address}` });
+    } else {
+      return <GrantRound>{};
+    }
+  });
+
+  // --- Contribution capabilities ---
+  const isContributing = ref(false);
+  const form = computed<{ token: string; amount: BigNumberish }>(() => {
+    return {
+      token: grantRound.value.donationToken,
+      amount: grantRound.value.minContribution,
+    };
+  });
+  const isFormValid = computed(() => {
+    const { token, amount } = form.value;
+    const areFieldsValid =
+      isValidAddress(<string>token) && token == grantRound.value.donationToken && (amount || 0 > 0);
+    return areFieldsValid;
+  });
+
+  /**
+   * @notice Check against the grantRounds status for a match
+   */
+  const hasStatus = (status: string) => (round: GrantRound) => round.status == status;
+
+  /**
+   * @notice Show the contribution window
+   */
+  function startContributing() {
+    isContributing.value = true; // show contribution form
+  }
+
+  /**
+   * @notice Hide the contribution window
+   */
+  function cancelContribution() {
+    isContributing.value = false; // hide contribution form
+  }
+
+  /**
+   * @notice Send contribution
+   */
+  async function sendContribution() {
+    // Get contract instances
+    if (!signer.value) throw new Error('Please connect a wallet');
+
+    // pull data from form
+    const { amount } = form.value;
+
+    // set up contracts
+    const token = new Contract(grantRound.value.donationToken, ERC20_ABI, signer.value);
+    const round = <GrantRoundContract>new Contract(grantRound.value.address, GRANT_ROUND_ABI, signer.value);
+
+    // contributionAmount must have the right number of decimals and be hexed
+    const contributionAmount = utils.parseUnits(amount.toString(), grantRound.value.donationTokenDecimals);
+
+    // check if contract is already approved as a spender
+    const allowance = await checkAllowance(token, grantRound.value.address);
+    if (allowance < contributionAmount) {
+      await getApproval(token, grantRound.value.address, contributionAmount);
```

Comment:
My personal preference is for approval amount to be `ethers.constants.MaxUint256` since contracts won't be upgradeable, so this will save user's gas in the long run

Reply:
Resolved in 85d0d88